x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/action/TransportStartDataFrameAnalyticsAction.java

```diff
@@ -0,0 +1,187 @@
+                            "] because it has already been opened", RestStatus.CONFLICT, e);
+                    }
+                    listener.onFailure(e);
+                }
+            };
+
+        // Start persistent task
+        ActionListener<DataFrameAnalyticsConfig> validatedConfigListener = ActionListener.wrap(
+            config -> persistentTasksService.sendStartRequest(MlTasks.dataFrameAnalyticsTaskId(request.getId()),
+                MlTasks.DATA_FRAME_ANALYTICS_TASK_NAME, taskParams, waitForAnalyticsToStart),
+            listener::onFailure
+        );
+
+        // Validate config
+        ActionListener<DataFrameAnalyticsConfig> configListener = ActionListener.wrap(
+            config -> DataFrameDataExtractorFactory.create(client, Collections.emptyMap(), config.getId(), config.getSource(),
+                ActionListener.wrap(dataExtractorFactory -> validatedConfigListener.onResponse(config), listener::onFailure)),
+            listener::onFailure
+        );
+
+        // Get config
+        configProvider.get(request.getId(), configListener);
+    }
+
+    public static class DataFrameAnalyticsTask extends AllocatedPersistentTask {
+
+        private final StartDataFrameAnalyticsAction.TaskParams taskParams;
+
+        public DataFrameAnalyticsTask(long id, String type, String action, TaskId parentTask, Map<String, String> headers,
+                                      StartDataFrameAnalyticsAction.TaskParams taskParams) {
+            super(id, type, action, "data_frame_analytics-" + taskParams.getId(), parentTask, headers);
+            this.taskParams = Objects.requireNonNull(taskParams);
+        }
+
+        public StartDataFrameAnalyticsAction.TaskParams getParams() {
+            return taskParams;
+        }
+    }
+
+    public static class TaskExecutor extends PersistentTasksExecutor<StartDataFrameAnalyticsAction.TaskParams> {
+
+        private final DataFrameAnalyticsManager manager;
+
+        public TaskExecutor(DataFrameAnalyticsManager manager) {
+            super(MlTasks.DATA_FRAME_ANALYTICS_TASK_NAME, MachineLearning.UTILITY_THREAD_POOL_NAME);
+            this.manager = Objects.requireNonNull(manager);
+        }
+
+        @Override
+        protected AllocatedPersistentTask createTask(
+            long id, String type, String action, TaskId parentTaskId,
+            PersistentTasksCustomMetaData.PersistentTask<StartDataFrameAnalyticsAction.TaskParams> persistentTask,
+            Map<String, String> headers) {
+            return new DataFrameAnalyticsTask(id, type, action, parentTaskId, headers, persistentTask.getParams());
+        }
+
+        @Override
+        protected DiscoveryNode selectLeastLoadedNode(ClusterState clusterState, Predicate<DiscoveryNode> selector) {
+            // For starters, let's just select the least loaded ML node
+            // TODO implement memory-based load balancing
+            return super.selectLeastLoadedNode(clusterState, MachineLearning::isMlNode);
```

Comment:
We need to make sure that in the future we remember to add these new tasks to our `upgrade_mode` settings. So that when users upgrade from 7.x -> 8.x they can do so without stopping these tasks themselves.